modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4TrashingAllocatorIT.java

```diff
@@ -0,0 +1,114 @@
+import org.elasticsearch.common.settings.SettingsFilter;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.plugins.ActionPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestHandler;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class Netty4TrashingAllocatorIT extends ESNetty4IntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.concatLists(List.of(Handler.class), super.nodePlugins());
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false;
+    }
+
+    public void testTrashContent() throws InterruptedException {
+        try (var client = new Netty4HttpClient()) {
+            var addr = randomFrom(internalCluster().getInstance(HttpServerTransport.class).boundAddress().boundAddresses()).address();
+            var responses = client.post(addr, List.of(new Tuple<>(Handler.ROUTE, randomAlphaOfLength(between(1024, 2048)))));
+            assertEquals(HttpResponseStatus.OK, responses.stream().findFirst().get().status());
+        }
+    }
+
+    public static class Handler extends Plugin implements ActionPlugin {
+        static final String ROUTE = "/_test/trashing-alloc";
+
+        @Override
+        public Collection<RestHandler> getRestHandlers(
+            Settings settings,
+            NamedWriteableRegistry namedWriteableRegistry,
+            RestController restController,
+            ClusterSettings clusterSettings,
+            IndexScopedSettings indexScopedSettings,
+            SettingsFilter settingsFilter,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            Supplier<DiscoveryNodes> nodesInCluster,
+            Predicate<NodeFeature> clusterSupportsFeature
+        ) {
+            return List.of(new BaseRestHandler() {
+                @Override
+                public String getName() {
+                    return ROUTE;
+                }
+
+                @Override
+                public List<Route> routes() {
+                    return List.of(new Route(RestRequest.Method.POST, ROUTE));
+                }
+
+                @Override
+                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+                    var iter = request.releasableContent().iterator();
+                    return (chan) -> {
+                        request.getHttpRequest().release();
```

Comment:
Are we sure that this releases the last ref? ISTR we thought there were other bits of Netty that might still hold a ref at this point, to be released when this stack of methods eventually returns.

Can we assert that the buffers are released?

Reply:
Will add assertion. I'm pretty sure I saw refCnt 2 before, but not anymore 🤥. The buffer is released indeed because following code asserts it's zeroed. Tried to switch allocator back to non-trashing still same refCnt=1. 

I do want to address acquire/release flow, as you mentioned before. Move release to netty stack and remove releases from the server package, leave only explicit retains/releases.

Reply:
> I do want to address acquire/release flow, as you mentioned before.

In a follow-up right?

Reply:
Yes